### PR TITLE
[plugin.audio.pureradio] 3.0.0

### PR DESCRIPTION
**Pull request by bdommel**
Rewrite based on guidelines for addons for Kodi 19+

### CODE REVIEW

File: plugin.audio.pureradio/resources/lib/plugin.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            link_item.setLabel2(item.get('label', item.get('info').get('title')))
+            link_item.setInfo(type = self.plugin_type, infoLabels = item.get('info'))
+            item_list.append((item.get('path'), link_item, not item.get('IsPlayable', False)))
+        xbmcplugin.addDirectoryItems(self.pluginhandle, item_list, len(item_list))
+        xbmcplugin.endOfDirectory(self.pluginhandle, succeeded=True, updateListing=updateListing, cacheToDisc=True)
+
+    def showWarning(self, message):
+        """Shows a popup window in the XBMC GUI for 5 seconds"""
+        self._log("showWarning message: %s" % message)
+        xbmc.executebuiltin('XBMC.Notification(Info:,' + message + ',6000)')
```
Comment (learningit) [inline plugin.audio.pureradio/resources/lib/plugin.py]: This is deprecated. It's better to  use xbmcgui.Dialog().notification()

File: plugin.audio.pureradio/resources/lib/addon.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+def run():
+    """This function is the entry point to the add-on.
+    It gets the add-on parameters and call the 'action' function.
+    """
+    p.log("pureradio.run")
+    # Get the params
+    params = p.get_plugin_parms()
+    action = params.get("action", '')
+    if action:
+        eval("%s(params)" % action)
```
Comment (enen92) [inline plugin.audio.pureradio/resources/lib/addon.py]: You should avoid using eval calls, especially in entry points : https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html

File: plugin.audio.pureradio/resources/lib/api.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import urllib.request as request
+from xml.etree import ElementTree as ET
+
+# functions with web interface
+def retrieve_streamurl():
+    return """http://pureradio.one:8000/"""
+
+def retrieve_podcasturl():
+    return """http://www.spreaker.com/show/1757189/episodes/feed"""
```
Comment (learningit) [inline plugin.audio.pureradio/resources/lib/api.py]: Are you sure this what you want here?

File: plugin.audio.pureradio/resources/lib/api.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# _*_ coding: utf-8 _*_
+
+import urllib.request as request
+from xml.etree import ElementTree as ET
+
+# functions with web interface
+def retrieve_streamurl():
+    return """http://pureradio.one:8000/"""
```
Comment (learningit) [inline plugin.audio.pureradio/resources/lib/api.py]: Are you sure about this return value?

File: plugin.audio.pureradio/resources/lib/plugin.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    def play_resolved_url(self, url = ""):
+        """Plays the media file pointed by the URL passed as argument."""
+        self._log("play_resolved_url pluginhandle = [%s] url = [%s]" % (self.pluginhandle, url))
+        listitem = xbmcgui.ListItem(path=url)
+        return xbmcplugin.setResolvedUrl(self.pluginhandle, True, listitem)
```
Comment (learningit) [inline plugin.audio.pureradio/resources/lib/plugin.py]: A general comment about some of your files. It's bad practice to not have an newline before the EOF.

**Comment by bdommel** [comment]
replaced eval with used static function calls 
replaced static urls with web response
replace depreciated notification with xbmcgui.Dialog().notification()
added newline at EOL

Thanks for the help!

File: plugin.audio.pureradio/resources/lib/addon.py (hunk 2 above)
Comment (bdommel) [reply plugin.audio.pureradio/resources/lib/addon.py]: replaced eval with static function calls

File: plugin.audio.pureradio/resources/lib/plugin.py (hunk 1 above)
Comment (bdommel) [reply plugin.audio.pureradio/resources/lib/plugin.py]: replaced the depricated executebuiltin function for xbmcgui.Dialog().notification()

File: plugin.audio.pureradio/resources/lib/api.py (hunk 4 above)
Comment (bdommel) [reply plugin.audio.pureradio/resources/lib/api.py]: replace the static string with response from web server.

File: plugin.audio.pureradio/resources/lib/api.py (hunk 3 above)
Comment (bdommel) [reply plugin.audio.pureradio/resources/lib/api.py]: replace the static string with response from web server.

File: plugin.audio.pureradio/resources/lib/plugin.py (hunk 5 above)
Comment (bdommel) [reply plugin.audio.pureradio/resources/lib/plugin.py]: added newline before EOF.

**Comment by bdommel** [comment]
The rebase did not go as expected. I will close this pull request and open a new one.

**Pull request closed without merging**